bootstore/src/schemes/v0/peer.rs

```diff
@@ -1217,80 +1419,60 @@ mod tests {
 
         // Wipe the learner ledger, restart the learner and instruct it to
         // relearn its share, and ensure that the neither generation number gets
-        // bumped because persistence doesn't occur.
-        learner_handle.shutdown().await.unwrap();
-        learner_jh.await.unwrap();
-        std::fs::remove_file(&learner_config.fsm_state_ledger_paths[0])
-            .unwrap();
-        let (mut learner, learner_handle) =
-            Node::new(learner_config.clone(), &log).await;
-        let learner_jh = tokio::spawn(async move {
-            learner.run().await;
-        });
-        let _ = learner_handle.load_peer_addresses(addrs.clone()).await;
-        learner_handle.init_learner().await.unwrap();
+        // bumped because persistence doesn't occur. But for that to happen
+        // we need to make sure the learner asks the same peer. The choice of
+        // which peer is somewhat implementation dependent, so to get around
+        // that we stop the other peer until after restarting the learner.
+
+        // Infer the correct peer based on which generation number didn't get
+        // bumped and stop it.
+        let non_learner_peer = if peer0_gen_new == peer0_gen { 0 } else { 1 };
+        nodes.shutdown_node(non_learner_peer).await;
+
+        // Now we can stop the learner, wipe its ledger, and restart it.
+        nodes.shutdown_learner(true).await;
+        nodes.start_learner().await;
+
+        // Now restart the other peer and tell the learner to relearn its share
+        nodes.start_node(non_learner_peer).await;
+        nodes[LEARNER].init_learner().await.unwrap();
+
+        // Ensure the peers' generation numbers didn't get bumped. The learner
+        // should've asked the same sled for a share first, which it already
+        // handed out.
         let peer0_gen_new_2 =
-            handle0.get_status().await.unwrap().fsm_ledger_generation;
+            nodes[0].get_status().await.unwrap().fsm_ledger_generation;
         let peer1_gen_new_2 =
-            handle1.get_status().await.unwrap().fsm_ledger_generation;
-
-        // Ensure the peer's generation numbers don't get bumped. The learner
-        // will ask the same sled for a share first, which it already handed
-        // out.
-        assert!(
-            peer0_gen_new == peer0_gen_new_2
-                && peer1_gen_new == peer1_gen_new_2
-        );
+            nodes[1].get_status().await.unwrap().fsm_ledger_generation;
+        assert_eq!(peer0_gen_new, peer0_gen_new_2);
+        assert_eq!(peer1_gen_new, peer1_gen_new_2);
```

Comment:
This probably warrants a close look to make sure the test is still valid. Without it, there was still some flakiness based since every time a node restarted it would have a new port and thus impact choice of peer.